### ISSUE TITLE
Handle crashing better

### ISSUE THREAD

**Issue by duymap** [ethereum-mining/ethminer#249]
Can we do the same mechanism like Claymore, ethminer should support a option like -r ( reboot ), in case there is something wrong during mining and can lead the app crashing, GPU stopped:
- If user configured option -r, if something crash ethminer should look for another .bat file to run
- the user will configure the reboot.bat which is same folder , used to reboot machine in case app crashed.
This would be helpful because we can configure to bootstrap ethiminer when PC rebooted. It means the app never stop for a while and better monitoring, if crashing happened, PC will auto reboot and re-run eth again. 

Thanks.

**Comment by smurfy** [comment]
I'm currently looking into ways to implement a simple watchdog function.
My ideas would be quite similar to claymore:
* Option to restart the miner application (not sure how exactly to do this)
* Option to exit the miner (the start script then can handle what to do)
* Option to execute a bat/shell script

I just need to figure out a way how to detect problems in the mining threads.

**Comment by AndreaLanfranchi** [comment]
As mentioned in another issue (sorry can't recall which one) I use this effective method.

1. Start ethminer automatically after a reboot. (this is an extract from my crontab)
`@reboot sleep 60 && bash ~/miner.sh`
2. Create a bash watchdog(ish) like this (I named it minerwd.sh):
```
# Requires  apt-get install inotify-tools
while inotifywait -e modify ~/miner.log > /dev/null 2>&1 ; do

  # Lookup last 10 rows of log file in search of errors
  # Feel free to integrate grep pattern or create more conditions
  if tail -n6 ~/miner.log | grep -io "cuda error\|error cuda"; then
  
	# Kill all instances of ethminer
	killall ethminer
	
	# Send mail
	echo "Miner requires restart due to error" | mail -s "Miner WatchDog Restart" prospector@localhost
    
	# Restart mining rig
	sudo /sbin/shutdown -f -r +2
	
	# Abandon WatchDog
	exit
	
  fi
done 
```
3. Have your miner.sh to launch both a bash watchdog and ethminer redirecting stdout and stderr to a log file.
```
# # Launch Ethminer watchdog
nohup ~/minerwd.sh & echo $! > ~/minerwd.pid
# # Launch Ethminer 
nohup ~/ethereum-mining/ethminer/build/ethminer/ethminer -U \
	-S "eth-eu1.nanopool.org:9999" -FS "eth-eu2.nanopool.org:9999" \
	-O "$ETH_ADDRESS.$RIGNAME/$MY_EMAIL" -v 5 --farm-recheck 500 \
	--dag-load-mode single 0 \
	>> ~/miner.log 2>&1 </dev/null & echo $! > ~/miner.pid

```
As you can see this gives you a quick and dirty way to restart the rig when an error occurs.
Change the search pattern for errors in log to add more conditions you might want to cause restart.

**Comment by ddobreff** [comment]
https://github.com/ddobreff/ethminer-manager
I have already done that and if GPU fails it reduces its consumption so based on that it autoreboots the rig with the option to send last 10 lines of screen to telegram channel. Be aware that if you are doing something else not hitting the power threashold limit and watchdog is active it will reboot without asking.

**Comment by smurfy** [comment]
Main "problem" with this solution they only work on linux :)  
I guess it makes sense to have a solution which works on windows too. i know there are watchdog tools for windows, but i mean one unified solution which works on all platforms the same way.

**Comment by AndreaLanfranchi** [comment]
Well ... these scripts are written for bash (linux) but, given the idea, it might be easy to implement the same exact path using Powershell on Windows.

**Comment by ddobreff** [comment]
It actually might be possible to use nvidia-smi.exe via powershell and check the power consumption in the way I do it on linux. As a roadmap an API will be very helpful for both operating systems.

**Comment by duymap** [comment]
Hi ,
Where I can find log file on window with ethminer ? I ran ethminer and did not found any .log file in the same folder with .exe file. I intend to write a small cron job to double check log file and it can work both in linux and window. thanks for your pointing out.

**Comment by AndreaLanfranchi** [comment]
Ethminer does not produce a log file.
It produces only screen output. If you want **that output** to be written to a file you have to use redirect pipes.

See here : https://www.lifewire.com/how-to-redirect-command-output-to-a-file-2618084

**Comment by Caerus7** [comment]
Some of us are using MinerLamp by orkblutt - https://github.com/orkblutt/MinerLamp

It's a nice solution until the devs get a watchdog into ethminer.  His latest version is supposed handle the per card reporting of the very latest 0.12 dev2 releases, but it just shows 0.00MH/s for me.  It works fine for the public and the published 0.12 dev2 releases though.

**Comment by duymap** [comment]
Hi,
I'm not a fan of watchdog. So I chose to implement a watcher using Java, here is first release : https://github.com/duymap/autoreboot-gpu-fail-mining, it based on gpu comsumption ( refer from https://github.com/ddobreff/ethminer-manager), so it only support for window auto reboot if GPU failed mining. Will support linux later. Appreciate for any idea/feedbacks.

**Comment by derubm** [comment]
windows - one of the multible solutions yet: >> https://github.com/derubm/Ethminer_Watchdog
basically two batchfiles and a windows port of tee to catch errors and restart the miinng process.
can be easily adapted with a little knowlegde of dos commands.

**Comment by ddobreff** [comment]
It is indeed very good solution but its still external.

**Comment by agbarbosa** [comment]
MinerLamp as per said by @Caerus7. It works as a charm for me in a NVidia environment.

**Comment by DeadManWalkingTO** [comment]
After #757 (added --exit parameter to exit whenever an error occurred) you can use a watchdog.

Try [ETHminerWatchDogDmW](https://github.com/DeadManWalkingTO/ETHminerWatchDogDmW) Windows7/8/10 [32/64] & Linux (Any Dist/Any Ver/Any Arch) (#735).

Check and feedback please.
Thank you!
